message_flag_updates: Fix handling of post failures for flag updates

kuv2707 opened the pull request:
If flag updation requests fail, we store the untracked changes in localStorage.
We make API calls to retry these changes on the next page load.
We optimize it by cancelling out add and remove calls for the same flag, and ignoring duplicate entries (like adding the same flag twice - though it currently isn't done by any of the callers and is included for the sake of completeness).


This PR builds up on #27858 but ended up diverging quite a bit.

[CZO](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Message.20not.20getting.20added.20to.20the.20Starred.20messages.2E)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 committed:
message_flags: Narrow down parameter type.

Judging by the use case, "op" can ever only be "add" or "remove".

ashutosh-engineer commented on web/src/message_flags.ts:
```diff
@@ -5,17 +5,135 @@ import * as z from "zod/mini";
+    cb: (entries: UntrackedFlagRecord) => UntrackedFlagRecord | undefined,
+): void {
+    if (localstorage.supported()) {
+        const ls = localstorage();
```
kindly recheck for the eslint issues and code structure

shuklamaneesh23 commented on web/tests/message_flags.test.cjs:
```diff
@@ -359,5 +365,143 @@ run_test("mark_as_unread", ({override}) => {
+        });
+    });
+
+    // ...This should work partially too.
```
Why `...` at the starting?

kuv2707 replied:
This is just to denote a train of thought.

roanster007 commented on web/src/message_flags.ts:
```diff
@@ -1,21 +1,139 @@
+export function all_local_untracked_flag_changes(
+    cb: (entries: UntrackedFlagRecord) => UntrackedFlagRecord | undefined,
+): void {
+    if (localstorage.supported()) {
```
just a small nitpick - probably better if there was an early return for `!localstorage.supported()` rather

aero31aero reviewed (changes requested):
I've taken a cursory look at this code, and did a bit of local testing. Testing didn't reveal any issues, as expected.

My feedback focuses more on code readability, and I've left 4-5 comments. Overall, its a complicated problem so the code is obviously complicated, but let's try to make this PR more readable. Nice work so far!

kuv2707 committed:
message_flags: Store failed flag update requests in localStorage.

If flag updation requests fail, we store the untracked changes in localStorage.
We make an API call to retry these changes on the next page load.

We try to save some API calls and reduce payload length
by using an optimisation where
we cancel out opposite changes to a message flag.

kuv2707 commented:
@zulipbot remove "mentor review"
@zulipbot add "maintainer review"

alya commented:
@timabbott would like to review this one directly.

timabbott commented on web/src/message_flags.ts:
```diff
@@ -1,11 +1,28 @@
 import * as z from "zod/mini";
 
 import * as channel from "./channel.ts";
+import {localstorage} from "./localstorage.ts";
```
I think the original version of this pull request didn't involve local storage, right? I am unsure if we want to do that; there's some tricky issues to manage involving multiple browser windows sharing the storage, and I worry it'll produce some nasty bugs. Where was the idea of making that change proposed? 

I agree there are some potential benefits, but I think it's fairly risky overall.

kuv2707 replied:
LocalStorage was used in #27858 so I also used it here.

I agree with your concern, I'll remove it.
